blockchain/blockchain.go

```diff
@@ -1045,23 +1074,56 @@ func (bc *BlockChain) writeStateTrie(block *types.Block, state *state.StateDB) e
 			trieDB.Commit(block.Header().Root, true, block.NumberU64())
 		}
 
-		if current := block.NumberU64(); current > triesInMemory {
-			// Find the next state trie we need to commit
-			header := bc.GetHeaderByNumber(current - triesInMemory)
-			chosen := header.Number.Uint64()
+		bc.chBlock <- gcBlock{root, block.NumberU64()}
+	}
+	return nil
+}
+
+// RLockGCCachedNode locks the GC lock of CachedNode.
+func (bc *BlockChain) RLockGCCachedNode() {
+	bc.stateCache.RLockGCCachedNode()
+}
+
+// RUnlockGCCachedNode unlocks the GC lock of CachedNode.
+func (bc *BlockChain) RUnlockGCCachedNode() {
+	bc.stateCache.RUnlockGCCachedNode()
+}
+
+// gcCachedNodeLoop runs a loop to gc.
+func (bc *BlockChain) gcCachedNodeLoop() {
+	trieDB := bc.stateCache.TrieDB()
 
-			// Garbage collect anything below our required write retention
-			for !bc.triegc.Empty() {
-				root, number := bc.triegc.Pop()
-				if uint64(-number) > chosen {
-					bc.triegc.Push(root, number)
-					break
+	bc.wg.Add(1)
+	go func() {
+		defer bc.wg.Done()
+		for {
+			select {
+			case block := <-bc.chBlock:
+				bc.triegc.Push(block.root, -float32(block.blockNum))
+				logger.Trace("Push GC block", "blkNum", block.blockNum, "hash", block.root.String())
+				blkNum := block.blockNum
+				if blkNum > triesInMemory {
```

Comment:
```suggestion
				if blkNum <= triesInMemory {
				    continue
				}
```
This will reduce an indentation 😅

Reply:
I updated. :)